### ISSUE TITLE
Update Databricks docs

### ISSUE THREAD

**Issue by AhdraMeraliQB** [kedro-org/kedro#3360]
## Description

It looks like Databricks have deprecated their cli tools, which has had the knock on effect of breaking our docs. A quick fix in #3358 adds the necessary `/archive/` to the broken links, but maybe we should rethink the section as a whole?

CC: @stichbury 

docs: https://docs.kedro.org/en/stable/deployment/databricks/databricks_ide_development_workflow.html

**Comment by astrojuanlu** [comment]
And also dbx:

> Databricks recommends that you use Databricks asset bundles instead of dbx by Databricks Labs. See [What are Databricks asset bundles?](https://docs.databricks.com/en/dev-tools/bundles/index.html) and [Migrate from dbx to bundles](https://docs.databricks.com/en/archive/dev-tools/dbx/dbx-migrate.html).

https://docs.databricks.com/en/archive/dev-tools/dbx/index.html

**Comment by stichbury** [comment]
I'll take a look at this in an upcoming sprint -- we did some updates for the asset bundles recently as suggested by Harmony.

**Comment by astrojuanlu** [comment]
Another couple of things I found in our Databricks workspace guide `databricks_notebooks_development_workflow.md`:

- The first half of the guide, which tells users to set up a GitHub repository, a personal access token, push the code, then create a Databricks Repo, is not strictly needed: `kedro new` works fine on Databricks notebooks.
  - One has to be careful that the files are created in the Workspace, and not in the ephemeral driver storage. Depending on the runtime version, the default working directory is different, so a `cd /Workspace/...` might be needed.
- "On Databricks, Kedro cannot access data stored directly in your project’s directory." this is not correct. From the docs:
  - "Spark cannot directly interact with workspace files on compute configured with shared access mode." however, clusters configured with Single User access mode should, and can, access workspace files.
  - However: “You can use Spark to read data files. You must provide Spark with the fully qualified path." https://learn.microsoft.com/en-us/azure/databricks/files/workspace-interact#read-data-workspace-files this means that `spark.read.load("/Workspace/...")` won't work (because it will assume `dbfs:/`), but `spark.read.load("file:/Workspace/...")` will.
    - Now, whether or not this can be incorporated in actual Kedro catalogs (in other words: whether or not our fsspec mangling will work on paths like these) is a different story. One can't just simply add `file:/` in front of the dataset filepath, because then it will be taken as an absolute path and not a relative one.

It's true that creating a Databricks Repo synced with a GH repo gives some nice advantages, like being able to edit the code in an actual IDE (whether a local editor or a cloud development environment like Gitpod or GitHub Codespaces). And it's also true that Databricks recommends in different places that data should live in the DBFS root.

However, it would be nice to consider what's the shortest and simplest guide we can write for users to _get started_ with Kedro on Databricks, and then build from there.

**Comment by astrojuanlu** [comment]
To clarify on the initial comment:

- The DBFS CLI (legacy) becomes https://docs.databricks.com/en/dev-tools/cli/fs-commands.html. The commands are still `databricks fs`, on first look nothing has changed on the CLI (only the structure of the documentation). For example, compare [old](https://docs.databricks.com/en/archive/dev-tools/cli/dbfs-cli.html#list-information-about-files-and-directories) and [new](https://docs.databricks.com/en/dev-tools/cli/fs-commands.html#list-the-contents-of-a-directory).
- The Jobs CLI (legacy) stops being documented because "CLI command groups that are not documented in the REST API reference have their own separate reference articles", and Jobs is not one of them. Documentation of the REST API can be found in https://docs.databricks.com/api/workspace/jobs, and the list of `jobs` subcommands is on GitHub at the moment https://github.com/databricks/cli/blob/main/docs/commands.md#databricks-jobs---manage-databricks-workflows. Less than ideal, but that's the current status AFAIU.

Both things above require the Databricks CLI version 0.205 and above. Apart from that, the commands haven't changed, so what we should do in this regard is making sure we're not sending users to legacy docs and that's it.

**Comment by astrojuanlu** [comment]
To summarise:

- [ ] Replace references to DBFS and Jobs legacy CLI docs
- [ ] Rewrite docs from dbx to Asset Bundles ([migration guide](https://docs.databricks.com/en/archive/dev-tools/dbx/dbx-migrate.html))
- [ ] Simplify Databricks notebooks guide to better serve as starting point for Kedro on Databricks

**Comment by stichbury** [comment]
@astrojuanlu Is this something that the team can pick up or do we need to ask for time from Jannic or another databricks expert (maybe @deepyaman could ultimately review)?

What are we prioritising this? I'm guessing it's relatively high importance to keep databricks docs tracking with their tech.

**Comment by astrojuanlu** [comment]
We'll need Databricks expertise in the team, so I hope we don't need to ask external experts to do it (it's OK if they give assistance, but we need to own this).

**Comment by astrojuanlu** [comment]
Added this to the Inbox so that we prioritise.

**Comment by deepyaman** [comment]
> @astrojuanlu Is this something that the team can pick up or do we need to ask for time from Jannic or another databricks expert (maybe @deepyaman could ultimately review)?

At this point, it's been almost 4 years since I've used Databricks (and don't currently have any interest in getting back into it), so I'd defer to somebody else. 🙂

**Comment by stichbury** [comment]
More than fair enough @deepyaman! Good to confirm though.

**Comment by astrojuanlu** [comment]
I'm adding one more item:

- [ ] Document integration with Databricks Unity Catalog

Every time I give a talk or workshop, invariably somebody from the audience asks "how does the Kedro Catalog play along with [Databricks Unity Catalog](https://www.databricks.com/product/unity-catalog)?".

Our reference docs for `kedro-datasets` mention it exactly once, in the API docs of [pandas.DeltaTableDataset](https://docs.kedro.org/projects/kedro-datasets/en/kedro-datasets-2.0.0/api/kedro_datasets.pandas.DeltaTableDataset.html).

And there's one subtle mention of it in [databricks.ManagedTableDataset](https://docs.kedro.org/projects/kedro-datasets/en/kedro-datasets-2.0.0/api/kedro_datasets.databricks.ManagedTableDataset.html) ("the name of the catalog in Unity".

The broader question of Delta datasets is a topic for https://github.com/kedro-org/kedro-plugins/issues/542.

**Comment by felipemonroy** [comment]
Maybe this could help:
 https://github.com/JenspederM/databricks-kedro-bundle

**Comment by astrojuanlu** [comment]
This looks really cool. @jenspederM do you want to share a bit more insight on how far do you intend to go with your project?

**Comment by JenspederM** [comment]
Hey @astrojuanlu 

Actually I don't really know if there's more to do. I almost want the project to be as barebone as possible. 

The way I left it now is with a very simple `datasets` implementation for unity so that people can customize as required. 

As for the DAB resource generator, I'm considering if I could find a better way for users to set defaults such as which job clusters, instance pools, etc..

Do you have any suggestions?

**Comment by astrojuanlu** [comment]
> However, it would be nice to consider what's the shortest and simplest guide we can write for users to _get started_ with Kedro on Databricks, and then build from there.

I gave two Kedro on Databricks demos yesterday, so I'm sharing that very simple notebook here https://github.com/astrojuanlu/kedro-databricks-demo hopefully it can be the basis of what I proposed in https://github.com/kedro-org/kedro/issues/3360#issuecomment-1836553608 (still no Kedro Framework there)

**Comment by astrojuanlu** [comment]
> Hey @astrojuanlu
> 
> Actually I don't really know if there's more to do. I almost want the project to be as barebone as possible.
> 
> The way I left it now is with a very simple `datasets` implementation for unity so that people can customize as required.
> 
> As for the DAB resource generator, I'm considering if I could find a better way for users to set defaults such as which job clusters, instance pools, etc..
> 
> One thing that is generally lacking is the documentation so that will definitely receive some attention once I have the time.
> 
> Do you have any suggestions?

@JenspederM I gave your `kedro-databricks` a quick try yesterday and it didn't work out of the box, so if you're open to me opening issues, I'll gladly start doing so 😄

**Comment by JenspederM** [comment]
@astrojuanlu Go for it!

I've been a bit busy these last few days and haven't had the chance to make any progress. 

But it's always nice to have some concrete issues to address. 😉

**Comment by JenspederM** [comment]
@astrojuanlu Just fyi, I'll merge a quite big PR soon so hopefully that will address most issues that you found. 

The substitution algorithm was a bit more cumbersome than first anticipated..

**Comment by astrojuanlu** [comment]
> Do we want to mention the kedro-databricks plugin as the shortest path to experiment Kedro on Databricks?

Yes!

> There is also a comment about UnityCatalog, is it in the scope of this ticket or should we separate this out?

It's in the scope of this ticket because this is a parent ticket. No need to do everything in the same PR.

The key thing is that we explain clearly how the Kedro Catalog and Unity Catalog can be used together. Example: https://github.com/astrojuanlu/kedro-databricks-demo/blob/main/First%20Steps%20with%20Kedro%20on%20Databricks.ipynb

> How do we think about the VSCode IDE Databricks

Do you mean https://docs.databricks.com/en/dev-tools/vscode-ext/index.html ? Not sure if there's anything Kedro-specific about that extension. We can discuss about that in another ticket.

**Comment by noklam** [comment]
@astrojuanlu 

> Do you mean https://docs.databricks.com/en/dev-tools/vscode-ext/index.html ? Not sure if there's anything Kedro-specific about that extension. We can discuss about that in another ticket.

Not directly, but it could be a better way to work with Databricks for the IDE experience. (no manual sync etc). Will keep this outside of the ticket.

**Comment by noklam** [comment]
## Steps to use `kedro-databricks`

1. Create starter `kedro new -s databrick-iris`
2. Install databricks-cli `curl -fsSL https://raw.githubusercontent.com/databricks/setup-cli/main/install.sh | sudo sh`
3. databricks configure copy link & create tokens
3. Install `kedro-databricks`
4. kedro databricks init #databricks.yml
5. kedro databricks bundle 
6. kedro databricks deploy
7. `databricks bundle run` or use the UI to trigger the job.


### Notes
> 2024-10-25 13:59:33,723 - databrick_iris - INFO - Substituting DBFS paths: Checking conf/base/catalog.yml
> 2024-10-25 13:59:33,724 - databrick_iris - INFO - Substituting DBFS paths: Checking conf/local/catalog.yml
> 2024-10-25 13:59:33,724 - databrick_iris - WARNING - Substituting DBFS paths: conf/local/catalog.yml does not exist.

When I run `kedro databricks init`, I see these logs showing up. I am a bit nervous about updating user file directly. For new file it's fine. This Create the `databricks.yml` that is needed 

## kedro databricks bundle
This create the resource/databricks-iris.yml,  wrapper around `databricks bundle init`

> Uploading databrick_iris-0.1-py3-none-any.whl...
Uploading bundle files to /Workspace/Users/nok_lam_chan@mckinsey.com/.bundle/databrick_iris/local/files...
Deploying resources...
Updating deployment state...
Deployment complete!


## kedro databricks deploy
wrapper around `databricks bundle deploy`

##  Changes
- The step to run a bundled job is not mentioned. Either through UI or using the `databricks` CLI directly with `databricks bundle run`

**Comment by JenspederM** [comment]
Yeah, I'm not too happy about having to change user paths either. 

But what is done, is simply to check that paths that refer to dbfs use the correct package name.

This is mostly a legacy feature that helped developer experience with the old version of the starter, where dbfs paths were not aligned with the package name. Since that has now changed, I might also remove this from the init command.

**Comment by noklam** [comment]
@DimedS Do you want to take the rest of the docs as I am working on the data asset bundles ?

**Comment by DimedS** [comment]
> @DimedS Do you want to take the rest of the docs as I am working on the data asset bundles ?

Yes, I’m ready to take care of them. I just need to wrap up the synthesis of our deployment interviews. Four of them were about Databricks, and we identified four distinct ways people deploy there. I believe we’ll complete the Databricks synthesis in the next two days. After that, I’ll share my thoughts on updating the Databricks docs here and will start making updates once we align on the changes.

**Comment by noklam** [comment]
Just want to share some notes for my DAB deployment experience with our own internal infra, there are two options: 1. Azure, 2. AWS instance

With Azure Databricks (the more common one), everything works well until the step 8 (run job with CLI) because we don't have job cluster with the sandbox.

With AWS Databricks, we do have job cluster, but I get stuck even earlier at  step 7 (`kedro databricks deploy`). After some investigation, it seems to related to permission issue. For some reason, I have permission to run a job, create a job through the UI but not with the CLI. I spent a good afternoon on it and decide it's not worth the effort to continue. This is the error that I see repeatly.

> Error: exit status 1
> Error: cannot read job: User nok_lam_chan@mckinsey.com does not have Admin or Manage Run or Owner or View permissions on job 1111051477632764
> 

To proceed, I will go with the Azure Databricks option, and use the UI to run the job as the last step. I expect users will likely encounter cryptic error message related to security/permission. I am not sure what kind of help we can provide other

**Comment by DimedS** [comment]
Hi, I'd like to share a synthesis of our recent interviews with four users who deploy projects to Databricks. In the following diagram, I’ve outlined their user flows, highlighting the main steps for deploying to Databricks. While there are many similarities, certain parts of the process differ.

<img width="1266" alt="Screenshot 2024-10-31 at 09 50 28" src="https://github.com/user-attachments/assets/74fc52e3-4e6a-45fe-a741-cc4fc523a753">

As you can see, the overall workflow is quite similar across users, but there are a few points where multiple options are available. I believe adding a diagram like this to our main Databricks documentation page (https://docs.kedro.org/en/stable/deployment/databricks/index.html) would be helpful. It could provide a concise overview of the steps required to deploy code and the various options for each step. This approach would be more efficient than detailing each complete workflow as we currently do. 

For instance, these two pages:
- https://docs.kedro.org/en/stable/deployment/databricks/databricks_ide_development_workflow.html
- https://docs.kedro.org/en/stable/deployment/databricks/databricks_notebooks_development_workflow.html

currently overlap by about 50%. Instead, we could focus on describing specific differences and options at each step, such as:

1. **Transferring Project Code to Databricks:**
   - Using the Databricks asset bundle
   - Manually syncing the repository
   - Using the VSCode plugin

2. **Running Kedro on Databricks:**
   - Via a packaged `.whl` file
   - Running code directly
   - Running code through a notebook

3. **Creating a Databricks Job to Schedule Runs:**
   - Setting up manually
   - Using the Databricks asset bundle
   - Using the Kedro-Databricks plugin (and its advantages)

4. **Choosing the Right Cluster for Jobs:**
   - **Job cluster**: Provides isolation but requires 5–10 minutes to start up
   - **Existing cluster**: Ready immediately but lacks isolation and can be costly to maintain 24/7 

This structure would allow users to better understand the options available at each step without redundant information.

**Comment by astrojuanlu** [comment]
This diagram is fantastic, and probably more useful than the current one we have.

Also +1 on trying to reduce overlap between the pages.

And finally, I noticed that some users didn't really want to `kedro package` their code, is that something you observed as well @DimedS ?

**Comment by DimedS** [comment]
> This diagram is fantastic, and probably more useful than the current one we have.
> 
> Also +1 on trying to reduce overlap between the pages.
> 
> And finally, I noticed that some users didn't really want to `kedro package` their code, is that something you observed as well @DimedS ?

Yes, from what I gathered, users didn’t find much benefit from packaging their projects - they only followed that approach because we recommended it. Some simplified the process and realised they could just upload code directly to their Databricks Repo, and they mentioned they don't need to package their projects. One user, Miguel, figured out that by specifying the notebook that runs the Kedro project within the Databricks Job deployment, he could use the same notebook during development, making debugging easier.

**Comment by noklam** [comment]
Do you plan to write these documentation, or it will be created as separate ticket from the current issue?

**Comment by astrojuanlu** [comment]
Another think we could try to document at some point is how to make use of the VS Code extension + `databricks-connect`. I tried and I got a `MissingConfigException: Given configuration path either does not exist or is not a valid directory: /Workspace/Users/juan_luis_cano@mckinsey.com/.bundle/kedro_databricks_playground/local/files/conf/local`

<img width="1624" alt="image" src="https://github.com/user-attachments/assets/61ea676e-aa48-4f77-a6f9-51b7111f7b7d">

And the workaround seems to be to do `git add -f conf/local` (see #2593).

**Comment by DimedS** [comment]
> Do you plan to write these documentation, or it will be created as separate ticket from the current issue?

I don’t think I proceed with it in the current sprint. Let’s create a new issue as a follow-up to the Databricks deployment research https://github.com/kedro-org/kedro/issues/4317. We can probably close current one after merging your PR, which transitions from dbx to Databricks asset bundles.

**Comment by astrojuanlu** [comment]
Current status of this issue:

- [x] P0: Deprecated `dbx` and documented Databricks Asset Bundles (#4265)
- [ ] P1: Describe more clearly how to connect to data _stored_ in Databricks
  - Source: User interviews November 2024
- [ ] P1: Write a very simple guide on how to use Kedro on Databricks without leaving the workspace (notebooks) (or simplify `databricks_notebooks_development_workflow.md` to the extreme)
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-1836553608
- [ ] P1: Document integration with Databricks Unity Catalog
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-1931660181
- [ ] P2: Fix links to the legacy DBFS CLI
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-1837284126
- [ ] P2: Document how to use the Databricks VS Code extension + `databricks-connect`
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-2470305411
- [ ] P2: Review claims on data storage ("On Databricks, Kedro cannot access data stored directly in your project’s directory")
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-1836553608
- [ ] P2: Rework decision diagram for deployment methods
  - Source: https://github.com/kedro-org/kedro/issues/3360#issuecomment-2449502744

**Comment by astrojuanlu** [comment]
> Azure Databricks rolled out new user interface elements that allow users to work directly with Git repo-backed folders from the Workspace UI, effectively replacing the prior, separate “Repos” feature functionality.

> If you are a user of the Databricks Repos feature for co-versioned Git-based source control of project assets, the core functionality has not changed. The most notable difference is that many contextual UI operations now refer to “Git folders” rather than “Repos”.

https://learn.microsoft.com/en-us/azure/databricks/repos/what-happened-repos

> Git folders can be created outside of the `/Repos` directory.
> Git folders can be created at various depths of the workspace file tree as long as they are under `/Workspace/Users/<user-email>` [...] You can have multiple Git folders under `/Workspace/Users/<user-email>`.
> you cannot create a new Git folder in Databricks without a remote repository URL.

**Comment by rashidakanchwala** [comment]
Review the current Databricks doc, remove outdated information and point to relevant documentation on Databricks or `kedro-databricks` since we don't manage those features.

**Comment by JenspederM** [comment]
`kedro-databricks` is also challenged at the moment by changes from Databricks. I'm tying to figure out how best to support the free version but there doesn't really seem to be any good solution as users' permissions on the platform vary wildly and there is no one size fits all. 

I would be honored by you just referring to the project but I need to resolve this. 

Suggestions are welcome!

**Comment by deepyaman** [comment]
Just throwing it out there that Ibis provides another option for working with Databricks. It circumvents anything related to Databricks Asset Bundles and Databricks Connect by just relying on the Databricks SQL client.

Of course, the main caveat is that you're writing Ibis code (as opposed to using the Spark API).

**Comment by JenspederM** [comment]
@deepyaman I think you misunderstand the problem. The problem is not interacting with databricks, it's creating the resources that allows you to interact with databricks. 

Since DBFS is deprecated (and has been for a while) the only real alternative is to use databricks volumes. These can be created through DAB but that would require that the user has the required permissions to do so. 

For now, DBFS only has soft deprecation, which means that it still works to upload data in the paid edition. However, in the free edition DBFS has been hard deprecated, I.e., is no longer supported. This means that either we will need to have two separate methods of deployment, one for the paid version and one for free version, but this comes with increased complexity both from a maintenance point of view and a usage point of view. 

The other solution is that we enforce the use of volumes and then if you don't have the required permissions to create one or you don't have one available to you - then you simply cannot use the plugin. 

If ibis has solved this problem, please provide a link to the documentation and I'll be happy to see if I can make it work for the plugin.

**Comment by deepyaman** [comment]
> [@deepyaman](https://github.com/deepyaman) I think you misunderstand the problem. The problem is not interacting with databricks, it's creating the resources that allows you to interact with databricks.

I don't think I misunderstand the problem, but I think there may be a broader miscommunication here.

The Databricks backend for Ibis is more akin to Databricks Connect in that it enables you to execute local code remotely, using Databricks as the execution engine. Since your code runs locally, there's no issue of packaging, unpackaging, uploading, bundling, etc. In that sense, it's simpler, and some of the resources don't need to be created; of course, there are caveats like non-Spark logic not being executed on the Databricks platform.

I think it's more relevant to the general question of "how to run Kedro pipelines on Databricks"  than fixing anything with the Kedro-Databricks workflow.

See also: [How to integrate Kedro and Databricks Connect](https://kedro.org/blog/how-to-integrate-kedro-and-databricks-connect)

**Comment by rashidakanchwala** [comment]
After discussing with @merelcht and @DimedS, we’ve decided to adjust the estimate for this ticket.
For now, it will be treated as a spike rather than a documentation update.

Databricks has evolved significantly, and instead of simply updating our existing (and somewhat outdated) docs, we want to take a broader view of how Kedro fits within the Azure Databricks ecosystem.

During this spike, we’ll explore:

- Which parts of the Databricks integration (e.g. Feature Store, etc.) are most relevant for Kedro.
- How Kedro can best leverage Databricks services and workflows.
- How to position and communicate a stronger, more comprehensive integration story through our documentation.

The goal of this sprint is to gather insights and outline the right direction.
Follow-up tickets will then focus on creating the actual documentation and examples based on what we learn.


cc - @Joseph-Perkins , @AliceCima10

**Comment by rashidakanchwala** [comment]
Closing this as we opened the summary ticket on next steps in #5209
